devito/propagator.py

```diff
@@ -4,20 +4,48 @@
 from sympy import symbols, IndexedBase, Indexed
 from function_manager import FunctionDescriptor
 from collections import Iterable
+from at_controller import get_at_block_size, get_optimal_block_size
 
 
 class Propagator(object):
-    def __init__(self, name, nt, shape, spc_border=0, forward=True, time_order=0, openmp=False, profile=False, cache_blocking=False, block_size=5):
+    def __init__(self, name, nt, shape, spc_border=0, forward=True, time_order=0, openmp=False, profile=False,
+                 cache_blocking=False, block_size=None, auto_tune=False):
         self.t = symbols("t")
+
         self.cache_blocking = cache_blocking
-        if(isinstance(block_size, Iterable)):
-            if(len(block_size) == len(shape)):
-                self.block_sizes = block_size
+        self.auto_tune = auto_tune
+
+        self.spc_order = spc_border * 2  # setting space order
+        self.time_order = time_order
+
+        if self.cache_blocking:
+            if self.auto_tune:  # if auto tuning get block optimal block sizes and tune around that
+                self.tune_b_size = get_optimal_block_size(shape, self.time_order, self.spc_order)
+
+                self.tune_range = 8  # range of tuning around tune_b_size
+                if self.tune_b_size - self.tune_range <= 0:  # making sure that tune_b - tune_range > 0
+                    self.tune_range -= - (self.tune_b_size - self.tune_range) + 1
+
+                block_size = self.tune_b_size
+            elif not block_size:
+                # else if block_size not set check if there is best block_size from at report else use optimal on
+                optimal_block_size = get_optimal_block_size(shape, self.time_order, self.spc_order)
+                block_size = get_at_block_size(self.time_order, self.spc_order)
```

Comment:
Because doing what the user wants us to do is more important than doing the "best". When the user sets auto_tune to off, most people aren't expecting you to use the results of a previous autotuning run.